clusterloader2/pkg/measurement/common/ooms_tracker.go

```diff
@@ -0,0 +1,200 @@
+	isRunning bool
+	stopCh    chan struct{}
+	ooms      []oomEvent
+}
+
+type oomEvent struct {
+	Node          string    `json:"node"`
+	Process       string    `json:"process"`
+	ProcessMemory string    `json:"memory"`
+	ProcessID     int       `json:"pid"`
+	Time          time.Time `json:"time"`
+}
+
+func (m *clusterOomsTrackerMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	clusterOomsTrackerEnabled, err := util.GetBoolOrDefault(config.Params, clusterOomsTrackerEnabledFlagName, false)
+	if err != nil {
+		return nil, err
+	}
+	if !clusterOomsTrackerEnabled {
+		klog.Info("skipping tracking of OOMs in the cluster")
+		return []measurement.Summary{}, nil
+	}
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		return []measurement.Summary{}, m.start(config.ClusterFramework.GetClientSets().GetClient())
+	case "gather":
+		summary, err := m.gather()
+		if err != nil {
+			return summary, err
+		}
+		return summary, m.validate()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) Dispose() {
+	m.stop()
+}
+
+func (m *clusterOomsTrackerMeasurement) String() string {
+	return clusterOomsTrackerName
+}
+
+func (m *clusterOomsTrackerMeasurement) start(c clientset.Interface) error {
+	if m.isRunning {
+		klog.Infof("%s: cluster OOMs tracking measurement already running", m)
+		return nil
+	}
+	klog.Infof("%s: starting cluster OOMs tracking measurement...", m)
+	m.isRunning = true
+	m.stopCh = make(chan struct{})
+	m.ooms = make([]oomEvent, 0)
+	// Watching for OOM events from node-problem-detector below.
+	i := informer.NewInformer(
+		c,
+		"events",
+		m.selector,
+		m.handleOomEvent,
+	)
+	return informer.StartAndSync(i, m.stopCh, informerTimeout)
+}
+
+func (m *clusterOomsTrackerMeasurement) stop() {
+	if m.isRunning {
+		m.isRunning = false
+		close(m.stopCh)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering cluster OOMs tracking measurement", clusterOomsTrackerName)
+	if !m.isRunning {
+		return nil, fmt.Errorf("measurement %s has not been started", clusterOomsTrackerName)
+	}
+
+	m.stop()
+
+	content, err := util.PrettyPrintJSON(struct {
+		Ooms []oomEvent `json:"ooms"`
+	}{
+		Ooms: m.ooms,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	summary := measurement.CreateSummary(clusterOomsTrackerName, "json", content)
+	return []measurement.Summary{summary}, nil
+}
+
+func (m *clusterOomsTrackerMeasurement) validate() error {
+	if len(m.ooms) == 0 {
+		return nil
+	}
+	return fmt.Errorf("OOMs recorded: %+v", m.ooms)
+}
+
+func (m *clusterOomsTrackerMeasurement) handleOomEvent(_, obj interface{}) {
+	if obj == nil {
+		return
+	}
+	event, ok := obj.(*corev1.Event)
+	if !ok {
+		return
+	}
+	oom := oomEvent{
+		Node: event.InvolvedObject.Name,
+		Time: event.FirstTimestamp.Time,
```

Comment:
Apparently, OOM events do not have EventTime field set, unfortunately. To emit events, `node-problem-detector` launches this part of k8sexporter [here](https://github.com/kubernetes/node-problem-detector/blob/452818cef8be2858f3c37f884b3583ba2463846a/pkg/exporters/k8sexporter/k8s_exporter.go#L72) which (indirectly) uses [this](https://github.com/kubernetes/client-go/blob/be97aaa976ad58026e66cd9af5aaf1b006081f09/tools/record/event.go#L347) function to create an event without `EventTime` field being set. Past manual tests of this change also notice only events with empty `EventTime` field as well.

Reply:
EventTime is used be the new API, FirstTimestamp is the old API one.
This will change in the future once we migrate.

You may want to do the same what we did in pod-startup SLO to address both at the same time:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/slos/pod_startup_latency.go#L192-L196